### DISCUSSION TITLE
Guided Assignment - React App with API Data Fetching - Magical Spellbook

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#173] [Tasks]
Refer to previous task #171 

Let's create a new and more interactive task called "Magical Spellbook" where users can create their own magical spells. They can add, edit, and delete spells, and each spell will have a name, description, and a unique magical effect. This exercise will involve more user interaction, state management, and a touch of whimsical fun. Here are the detailed step-by-step instructions:

Step 1: Setup
- Create a new directory for your project.
- Initialize a new React app using Create React App by running the command: `npx create-react-app magical-spellbook`.
- Navigate into the project directory: `cd magical-spellbook`.
- Start the development server: `npm start`.
- Open your browser and go to `http://localhost:3000` to see the default React app running.

Step 2: Clear the default project
- Open the `src/App.js` file and remove the existing code.
- Replace it with the following code:
```jsx
import React, { useState } from 'react';
import SpellForm from './SpellForm';
import SpellList from './SpellList';

function App() {
  const [spells, setSpells] = useState([]);

  const addSpell = (spell) => {
    setSpells((prevSpells) => [...prevSpells, spell]);
  };

  const updateSpell = (updatedSpell) => {
    setSpells((prevSpells) =>
      prevSpells.map((spell) => (spell.id === updatedSpell.id ? updatedSpell : spell))
    );
  };

  const deleteSpell = (spellId) => {
    setSpells((prevSpells) => prevSpells.filter((spell) => spell.id !== spellId));
  };

  return (
    <div className="container">
      <h1 className="text-center">Magical Spellbook</h1>
      <SpellForm addSpell={addSpell} />
      <SpellList spells={spells} updateSpell={updateSpell} deleteSpell={deleteSpell} />
    </div>
  );
}

export default App;
```

Step 3: Create SpellForm component
- Create a new file called `SpellForm.js` in the `src` directory.
- Open `SpellForm.js` and add the following code:
```jsx
import React, { useState } from 'react';

function SpellForm({ addSpell }) {
  const [spell, setSpell] = useState({ name: '', description: '', effect: '' });

  const handleInputChange = (event) => {
    const { name, value } = event.target;
    setSpell((prevSpell) => ({ ...prevSpell, [name]: value }));
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (spell.name.trim() !== '' && spell.description.trim() !== '' && spell.effect.trim() !== '') {
      addSpell({ ...spell, id: Date.now() });
      setSpell({ name: '', description: '', effect: '' });
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <div className="mb-3">
        <label htmlFor="name" className="form-label">
          Name
        </label>
        <input
          type="text"
          name="name"
          id="name"
          className="form-control"
          value={spell.name}
          onChange={handleInputChange}
        />
      </div>
      <div className="mb-3">
        <label htmlFor="description" className="form-label">
          Description
        </label>
        <textarea
          name="description"
          id="description"
          className="form-control"
          rows="3"
          value={spell.description}
          onChange={handleInputChange}
        ></textarea>
      </div>
      <div className="mb-3">
        <label htmlFor="effect" className="form-label">
          Effect
        </label>
        <input
          type="text"
          name="effect"
          id="effect"
          className="form-control"
          value={spell.effect}
          onChange={handleInputChange}
        />
      </div>
      <button type="submit" className="btn btn-primary">
        Add Spell
      </button>
    </form>
  );
}

export default SpellForm;
```

Step 4: Create SpellList component
- Create a new file called `SpellList.js` in the `src` directory.
- Open `SpellList.js` and add the following code:
```jsx
import React from 'react';

function SpellList({ spells, updateSpell, deleteSpell }) {
  return (
    <ul className="list-group mt-4">
      {spells.map((spell) => (
        <li key={spell.id} className="list-group-item">
          <div className="d-flex justify-content-between">
            <div>
              <h5 className="mb-0">{spell.name}</h5>
              <small className="text-muted">{spell.description}</small>
            </div>
            <div>
              <button
                className="btn btn-sm btn-warning me-2"
                onClick={() => updateSpell({ ...spell, effect: 'Magic!' })}
              >
                Cast Spell
              </button>
              <button
                className="btn btn-sm btn-danger"
                onClick={() => deleteSpell(spell.id)}
              >
                Delete
              </button>
            </div>
          </div>
          <p className="mt-2 mb-0">Effect: {spell.effect}</p>
        </li>
      ))}
    </ul>
  );
}

export

default SpellList;
```

Step 5: Add Bootstrap styles
- Open the `src/index.js` file.
- Replace the existing code with the following code:
```jsx
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import 'bootstrap/dist/css/bootstrap.min.css';
import './index.css';

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root')
);
```

Step 6: Create custom CSS
- Create a new file called `index.css` in the `src` directory.
- Open `index.css` and add the following code:
```css
body {
  background-color: #f4f4f4;
  padding-top: 2rem;
}

.container {
  max-width: 600px;
}

.form-control,
.btn {
  margin-bottom: 1rem;
}

.list-group-item {
  display: flex;
  flex-direction: column;
}

.list-group-item .d-flex {
  align-items: center;
}

.list-group-item .d-flex .btn {
  margin-top: 0.5rem;
}

.list-group-item .mt-2 {
  font-size: 0.875rem;
  color: #555;
}
```

Step 7: Run the app
- Make sure the development server is running (`npm start`).
- Open your browser and go to `http://localhost:3000`.
- You should see the "Magical Spellbook" heading and a form to add spells.
- Fill in the spell name, description, and effect, then click "Add Spell" to add it to the spell list.
- The added spell will appear in the list below.
- Each spell has options to cast the spell (which changes the effect) and delete the spell.

Congratulations! You have created the "Magical Spellbook" React app. Users can add, edit, and delete magical spells. The app utilizes useState for managing spells and their state. Bootstrap and custom CSS provide a visually appealing and interactive user interface. Feel free to further customize the styles, add additional features, or incorporate more magical elements to make it even more whimsical and engaging. Enjoy creating magical spells with your spellbook app!

**Comment by bipin-coded** [comment]
![Screenshot (5)](https://github.com/akash-coded/mern/assets/126976134/0a0675fc-1d00-465d-beeb-2e0d018d6670)

**Comment by chukku10** [comment]
![spellbook](https://github.com/akash-coded/mern/assets/127116126/f6eaa51a-7838-4952-bfc9-d7b870154dc5)

**Comment by SagarTChandrakanth** [comment]
![Screenshot from 2023-07-08 22-01-15](https://github.com/akash-coded/mern/assets/110445544/cdd6d735-8c5a-4233-9e5f-0f5dc1d3f7e8)

**Comment by Karthikkevin** [comment]
![Screenshot (156)](https://github.com/akash-coded/mern/assets/33948674/40712979-0c8d-488c-8614-033a594d5cb3)

**Comment by Sheetal-dh** [comment]
![image](https://github.com/akash-coded/mern/assets/91683804/67015719-f20d-4792-ad4a-145a769f89ee)

**Comment by Praveen02062000** [comment]
![Screenshot (4)](https://github.com/akash-coded/mern/assets/137034312/841bfac2-aca9-4bc8-8772-297f20a23e45)

**Comment by pooja744** [comment]
![image](https://github.com/akash-coded/mern/assets/138996416/5ed19baa-3bd2-4d35-84a9-d1575f66bd9f)

**Comment by rishab1616** [comment]
![image](https://github.com/akash-coded/mern/assets/52340386/3870984b-e61b-4b4b-861c-f4d3083dd5e5)

**Comment by priyamallireddy** [comment]
<img width="721" alt="Screenshot 2023-07-08 221232" src="https://github.com/akash-coded/mern/assets/137372978/100500b5-95d1-4067-b8d5-fa386a9be3ba">

**Comment by nandini-naik09** [comment]
![image](https://github.com/akash-coded/mern/assets/133530727/099fa0ad-f2e6-4b70-987a-fab52c18e902)

**Comment by Ruchiarth** [comment]
![image](https://github.com/akash-coded/mern/assets/126975575/6203954a-7876-4571-a24d-2cc6d3e5e3ed)
![image](https://github.com/akash-coded/mern/assets/126975575/370b3fff-8c9c-4b3f-8dbe-ff39b0d25d43)

**Comment by madhuriS3008** [comment]
![image](https://github.com/akash-coded/mern/assets/85974647/b79f62e9-191f-44fa-9d70-7614c74f409b)

**Comment by mownica-p** [comment]
![image](https://github.com/akash-coded/mern/assets/126975660/a93a4e69-56e7-4242-ba09-79ed7f9a7b08)

**Comment by Ravi-710** [comment]
![Screenshot (86)](https://github.com/akash-coded/mern/assets/98272814/26a3a34a-3b8b-49bf-98de-42bce0f537a6)

**Comment by KavyaKetha** [comment]
**$ npm install axios bootstrap --save** is not specified in the steps it will throw bootstrap css file error.
![image](https://github.com/akash-coded/mern/assets/12147689/29331278-42ef-4815-9897-b82651c36d39)

**Comment by Sanket866** [comment]
<img width="438" alt="image" src="https://github.com/akash-coded/mern/assets/120538359/33310601-a1ff-4682-a77d-54b9bbdbe919">
.

**Comment by pawarprathmesh333** [comment]
![image](https://github.com/akash-coded/mern/assets/71115267/49db04c1-5ae2-4c55-a53a-7cf30e4cfea1)

**Comment by vk0027** [comment]
![image](https://github.com/akash-coded/mern/assets/123943655/2956ee89-849e-4d34-8864-94f12a0a68c5)

**Comment by starvikas1988** [comment]
![image](https://github.com/akash-coded/mern/assets/70792693/48b99460-4cf3-4003-87e3-7a93c5643b87)

**Comment by RevathiSundaramoorthy** [comment]
![image](https://github.com/akash-coded/mern/assets/126975777/7a4f80e4-c611-4cf0-a54b-acfe596f4e02)

**Comment by sam2elect** [comment]
![image](https://github.com/akash-coded/mern/assets/85131631/d9873f07-b5c2-4c7c-a1dc-339293bb0ba2)
